File: backend/src/App.Core/Entities/Transfer.cs
namespace App.Core.Entities;

/// <summary>
/// 调拨单实体（对应 PostgreSQL 表 Transfers，specs/039-erp-transfer）。
/// 一步式单据：保存即生效（转出仓 −、转入仓 +，同一事务）；不支持编辑，只支持作废回冲。
/// 转出 / 转入仓名称、明细商品编码 / 名称 / 单位均为快照，后续档案修改不影响历史单据。
/// 调拨不落业务金额（无价格概念）；列表只展示「数量合计」与「明细行数」。
/// </summary>
public sealed class Transfer
{
    /// <summary>调拨单 ID</summary>
    public Guid Id { get; set; }

    /// <summary>单号，唯一，后端生成（TR + yyyyMMdd + 4 位序号，如 TR202609170001）</summary>
    public string TransferNo { get; set; } = string.Empty;

    /// <summary>转出仓 ID（外键 → Warehouses(Id)，038）</summary>
    public Guid FromWarehouseId { get; set; }

    /// <summary>转出仓名称快照（列表 / 详情 / 导出免 join；仓改名后历史单据保持当时名称）</summary>
    public string FromWarehouseName { get; set; } = string.Empty;

    /// <summary>转入仓 ID（外键 → Warehouses(Id)，038）</summary>
    public Guid ToWarehouseId { get; set; }

    /// <summary>转入仓名称快照（列表 / 详情 / 导出免 join）</summary>
    public string ToWarehouseName { get; set; } = string.Empty;

    /// <summary>业务日期（UTC 午夜）</summary>
    public DateTimeOffset TransferDate { get; set; }

    /// <summary>明细行数（列表展示，后端重算）</summary>
    public int ItemCount { get; set; }

    /// <summary>数量合计（列表展示，后端重算）</summary>
    public int TotalQuantity { get; set; }

    /// <summary>单据状态（1=正常 0=已作废；作废后禁止再操作）</summary>
    public OrderStatus Status { get; set; } = OrderStatus.Normal;

    /// <summary>备注</summary>
    public string? Remark { get; set; }

    /// <summary>创建时间</summary>
    public DateTimeOffset CreatedAt { get; set; }

    /// <summary>更新时间</summary>
    public DateTimeOffset UpdatedAt { get; set; }

    /// <summary>创建人用户 id</summary>
    public Guid? CreatedBy { get; set; }

    /// <summary>更新人用户 id</summary>
    public Guid? UpdatedBy { get; set; }
}

File: frontend/src/api/transfer.ts
import type { PagedResult } from './product'
import { get, post, put } from './request'

/** 调拨单状态（0 已作废 / 1 正常，复用后端 OrderStatus 整型口径） */
export type TransferStatus = 0 | 1

/** 调拨单列表行（对应后端 TransferListItemDto；快照仓名原样返回，免联查） */
export interface TransferListItem {
  id: string
  transferNo: string
  fromWarehouseId: string
  /** 转出仓名称快照（保存即固化） */
  fromWarehouseName: string
  toWarehouseId: string
  /** 转入仓名称快照（保存即固化） */
  toWarehouseName: string
  transferDate: string
  /** 商品行数（列表展示） */
  itemCount: number
  /** 数量合计（列表展示） */
  totalQuantity: number
  status: TransferStatus
  createdAt: string
}

/** 调拨单明细行（对应后端 TransferItemDto；040 前无 batchId，调拨无价格 → 无金额列） */
export interface TransferItem {
  id: string
  productId: string
  productCode: string
  productName: string
  unit: string
  quantity: number
}

/** 调拨单详情（对应后端 TransferDetailDto，明细按插入顺序） */
export interface TransferDetail extends Omit<TransferListItem, 'status'> {
  remark: string | null
  createdBy: string | null
  items: TransferItem[]
  status: TransferStatus
}

/** 调拨单列表查询参数（对应后端 GetTransfersRequest；转出 / 转入仓可空 = 全部仓） */
export interface TransferQuery {
  page: number
  pageSize: number
  keyword?: string
  fromWarehouseId?: string
  toWarehouseId?: string
  start?: string
  end?: string
}

/** 新增调拨单入参（对应后端 CreateTransferRequest；明细仅含 productId / quantity，调拨无价格） */
export interface CreateTransferPayload {
  fromWarehouseId: string
  toWarehouseId: string
  transferDate: string
  items: { productId: string; quantity: number }[]
  remark?: string
}

/**
 * 把页面选择的本地日期范围转换为后端所需的 UTC ISO 闭区间：
 * 起始取当天本地 00:00:00、结束取当天本地 23:59:59.999，再转 UTC（同采购 / 销售约定）。
 */
export function toDateRange(
  startDate?: string,
  endDate?: string,
): { start?: string; end?: string } {
  const start = startDate ? new Date(`${startDate}T00:00:00`).toISOString() : undefined
  const end = endDate ? new Date(`${endDate}T23:59:59.999`).toISOString() : undefined
  return { start, end }
}

/**
 * 把所选本地日期（YYYY-MM-DD）转换为后端所需的「所选日期的 UTC 午夜」ISO 串（design §4.2）。
 * 直接以 UTC 解释该日历日 00:00:00，避免后端把裸日期按服务器本地时区解析
 * （Npgsql 拒绝把非 UTC 偏移的 DateTimeOffset 写入 timestamptz 列）。
 */
export function toUtcMidnight(date: string): string {
  return new Date(`${date}T00:00:00Z`).toISOString()
}

/** 分页查询调拨单（含作废单据，作废行前端置灰） */
export function getTransfers(query: TransferQuery): Promise<PagedResult<TransferListItem>> {
  return get<PagedResult<TransferListItem>>('/transfers', { params: query })
}

/** 查询调拨单详情（含明细行，快照字段原样返回） */
export function getTransfer(id: string): Promise<TransferDetail> {
  return get<TransferDetail>(`/transfers/${id}`)
}

/** 新增调拨单（一步式：保存即生效，转出仓 −、转入仓 + 同一事务；返回详情含后端生成的单号与统计） */
export function createTransfer(payload: CreateTransferPayload): Promise<TransferDetail> {
  return post<TransferDetail>('/transfers', payload)
}

/** 作废调拨单（双向回冲库存与流水；仅改状态不删数据） */
export function voidTransfer(id: string): Promise<TransferDetail> {
  return put<TransferDetail>(`/transfers/${id}/void`)
}
